bsp/bsp.cpp

```diff
@@ -166,14 +166,22 @@ void ExecuteWakeProcess::execute() {
 
   pid_t child = fork();
   if (child == 0) {
-    // Redirect stdout to the pipe
-    for (int i = 0; i < PIPES; ++i) {
+    // Close the reading side of the pipe
+    for (int i = 0; i < PIPES; ++i)
       close(pipefds[i][0]);
-      // !!!
-      if (pipefds[i][1] != i+1) {
-        dup2(pipefds[i][1], i+1);
-        close(pipefds[i][1]);
-      }
+
+    // We need to dup2 pipefds[i][0,PIPES) to [1,PIPES]
+    // Unfortunately, some of the these might be in the way
+    // dup() any of them that are in the way, out of the way
+    for (int i = 0; i < PIPES; ++i)
+      while (pipefds[i][1] >= 1 && pipefds[i][1] <= PIPES)
+        pipefds[i][1] = dup(pipefds[i][1]);
+
+    // Put the pipes in their proper final position
+    // This also closes any transient descriptors from dup() above
+    for (int i = 0; i < PIPES; ++i) {
+      dup2(pipefds[i][1], i+1);
```

Comment:
dup and dup2 have errors that can be checked

Reply:
and close

since it's debugging we probably want all the info we can get

Reply:
While true, it's too late (post fork) to really do anything useful with them.
I guess I can print an error and not launch the subprocess.

Reply:
Actually, where would I even print it? This is what sets up stderr. :-/

Reply:
I'll just return rainbow-coloured exit codes.